_overviews/tutorials/binary-compatibility-for-library-authors.md

```diff
@@ -178,11 +178,13 @@ Again, we recommend using MiMa to double-check that you have not broken binary c
 Sometimes, it is desirable to change the definition of a case class (adding and/or removing fields) while still staying backwards-compatible with the existing usage of the case class, i.e. not breaking the so-called _binary compatibility_. The first question you should ask yourself is “do you need a _case_ class?” (as opposed to a regular class, which can be easier to evolve in a binary compatible way). A good reason for using a case class is when you need a structural implementation of `equals` and `hashCode`.
 
 To achieve that, follow this pattern:
- * make the primary constructor private (this makes the `copy` method of the class private as well)
- * define a private `unapply` function in the companion object (note that by doing that the case class loses the ability to be used as an extractor in match expressions)
+ * make the primary constructor private (for Scala 3, this makes the `copy` method of the class and `apply` in the companion object private as well)
  * for all the fields, define `withXXX` methods on the case class that create a new instance with the respective field changed (you can use the private `copy` method to implement them)
  * create a public constructor by defining an `apply` method in the companion object (it can use the private constructor)
- * in Scala 2, you have to add the compiler option `-Xsource:3`
+ * (Scala 2) you have to add the compiler option `-Xsource:3`
+ * (Scala 2) define the `copy` method with all the current fields manually and set it as `private`
+ * (Scala 2) define a private `unapply` method in the companion object (note that by doing that the case class loses the ability to be used as an extractor in match expressions)
```

Comment:
Why make this Scala 2 only? AFAIK it is also required in Scala 3.

Reply:
Not according to @lihaoyi 

> `unapply`: I think as of Scala 3 this will work right out of the box: `unapply` no longer returns `Option[TupleN[...]]` as it did in Scala 2, and instead just returns `Person`, with pattern matching just relying on the `._1` `._2` etc. fields to work. Thus, a `p match{ case Person(first, last) => ???}` callsite compiled against `case class Person(first: String, last: String)` should continue to work even when `Person` has evolved into `case class Person(first: String, last: String, country: String = "unknown", number: Option[String] = None)`

https://contributors.scala-lang.org/t/can-we-make-adding-a-parameter-with-a-default-value-binary-compatible/6132?u=sideeffffect

Reply:
That would be a source incompatibility, but if someone links an already compiled program with the new version of the case class, then they may get a run-time crash. I think it is preferable to recommend making `unapply` private.